### PR TITLE
Kmk

### PR DESCRIPTION
**Pull request by rohe**
Changed the behaviour when a signed JWT is received that contains no kid while the JWKS of the issuer contains more then one key that could have been used.

The default behaviour is that the verification should fail when this happens.
This is what the standard specifies as the expected behaviour.

But if one wants it's now possible to specify issuers that are 'allowed' to not include a kid.
For those issuers the possible set of keys will be checked one at the time.

### CODE REVIEW

File: tests/test_oauth2_message.py [hunk 1]
```diff
@@ -30,9 +30,35 @@
 from oic.oauth2.message import json_deserializer
 from oic.oauth2.message import json_serializer
 from oic.oauth2.message import sp_sep_list_deserializer
+from oic.utils.keyio import build_keyjar
 
 __author__ = 'rohe0002'
 
+keys = [
```
Comment (schlenk) [inline tests/test_oauth2_message.py]: All this setup should probably be in one or more pytest.fixture functions.

It clearly is a test fixture.

File: tests/test_oauth2_message.py [hunk 2]
```diff
@@ -595,3 +621,100 @@ def test_missing_required(self):
 
         with pytest.raises(MissingRequiredAttribute):
             err.to_urlencoded()
+
+
+def test_to_jwt_rsa():
```
Comment (schlenk) [inline tests/test_oauth2_message.py]: Those tests could maybe be simplified into just one function with https://docs.pytest.org/en/latest/parametrize.html#

The body is nearly the same, just a few different parameters.

File: src/oic/oauth2/message.py [hunk 3]
```diff
@@ -473,28 +473,55 @@ def to_jwt(self, key=None, algorithm="", lev=0):
...
+                        if allowed_kids:
+                            for k in kl:
+                                if k.kid in allowed_kids:
+                                    key.append(k)
+                        else:
+                            for k in kl:
+                                key.append(k)
 
     def get_verify_keys(self, keyjar, key, jso, header, jwt, **kwargs):
         try:
```
Comment (schlenk) [inline src/oic/oauth2/message.py]: docstring would be nice.

File: src/oic/oauth2/message.py [hunk 4]
```diff
@@ -473,28 +473,55 @@ def to_jwt(self, key=None, algorithm="", lev=0):
...
+                    if kl[0] not in key:
+                        key.append(kl[0])
+                elif no_kid_issuer:
+                    try:
+                        allowed_kids = no_kid_issuer[issuer]
+                    except KeyError:
+                        return
+                    else:
+                        if allowed_kids:
+                            for k in kl:
```
Comment (lwm) [inline src/oic/oauth2/message.py]: :+1 and maybe change k->key, k1->verified for future us.

**Comment by lwm** [comment]
Solid work. Please leave a change log entry too, should we consider a release after this?

**Comment by rohe** [comment]
Yes we should do a new release !

**Pull request merged**